### PR TITLE
Create first template according to demo

### PR DESCRIPTION
**Pull request by ElviaBth**
🎩 What? Why?
Create new organization from template page.

📌 Related Issues
#10 

📷 Screenshots
image
image
image

♥️ Thank you!

### CODE REVIEW

File: app/commands/decidim/system/create_template_fields.rb [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+module Decidim::System
+  # A command with all the business logic to create the default content blocks
+  # for a newly-created organization.
+  class CreateTemplateFields < Decidim::Command
+    DEFAULT_CONTENT_BLOCKS = [:hero, :footer_sub_hero, :highlighted_content_banner, :highlighted_consultations, :upcoming_meetings].freeze
```
Comment (microstudi) [inline app/commands/decidim/system/create_template_fields.rb:7]:
```suggestion
```
Not needed is it?

File: lib/templates/template-1.yml
Comment (microstudi) [inline lib/templates/template-1.yml:1]: Let's name this `template-consultations.yml`, also the subdirectory

File: lib/templates/template-1.yml [hunk 2]
```diff
@@ -3,16 +3,62 @@ name: Votacions AFFaC
...
+    weight: 1
+    scope_name: homepage
+    images_container:
+      - name: background_image
+        file: template-1/hero_background_image.png
+        content_type: image/png
+    settings:
+      welcome_text:
+        ca: Hola!
+        es: Hola!
```
Comment (microstudi) [inline lib/templates/template-1.yml:26]:
```suggestion
        ca: %{organization_name}
        es: %{organization_name}
```
probably something like that

File: app/commands/decidim/system/register_custom_templates.rb [hunk 3]
```diff
@@ -30,7 +30,7 @@ def call
           @organization = create_organization
           CreateDefaultPages.call(@organization)
```
Comment (microstudi) [inline app/commands/decidim/system/register_custom_templates.rb:31]:
```suggestion
          CreateCustomDefaultPages.call(@organization)
```
We will need our custom pages based on the template

**Review by microstudi — COMMENTED** [review]
4 notes!

File: app/commands/decidim/system/create_template_fields.rb (hunk 1 above)
Comment (ElviaBth) [reply app/commands/decidim/system/create_template_fields.rb:7]: I'll check. Thanks

File: lib/templates/template-1.yml (hunk 2 above)
Comment (ElviaBth) [reply lib/templates/template-1.yml:26]: Done.

File: app/commands/decidim/system/register_custom_templates.rb (hunk 3 above)
Comment (ElviaBth) [reply app/commands/decidim/system/register_custom_templates.rb:31]: Ok. Done.

File: lib/templates/template-1.yml
Comment (ElviaBth) [reply lib/templates/template-1.yml:1]: ok. Done.

**Comment by ElviaBth** [comment]
> 4 notes!

thanks. Done

File: lib/templates/template-consultations.yml [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            <li>No puedes dar a entender que %{organization_name} patrocina o respalda la actividad en la que se reutiliza la información.</li>
+            </ul><p>%{organization_name} respetará todos los derechos de propiedad intelectual o industrial sobre el contenido remitido por las personas usuarias, y no asume ninguna responsabilidad, directa o indirecta, respecto de cualquier tipo de controversia, disputa y/o litigio que pueda derivar de la publicación, divulgación y/o difusión de los contenidos aportados sin el consentimiento preceptivo de sus titulares legítimos o sin el amparo de algún uso permitido por la ley.</p></div>
+        weight: 1
+vars:
+  organization_name: translated_string
+  ref: string
+  host: string
+  organization_admin_name: string
+  organization_admin_email: string
+  year: string
```
Comment (microstudi) [inline lib/templates/template-consultations.yml:220]:
```suggestion
```

File: spec/system/organizations_spec.rb [hunk 5]
```diff
@@ -9,32 +9,93 @@
...
+
+        blocks = Decidim::ContentBlock
+        expect(blocks.count).to eq(5)
+        block_hero = blocks.find_by(manifest_name: :hero)
+        expect(block_hero.settings.welcome_text).to eq({ "ca" => "Benvinguda Citizen Corp", "es" => "Bienvenida Citizen Corp" })
+        expect(block_hero.images_container.background_image.attached?).to be true
+        expect(block_hero.images_container.attached_uploader(:background_image).path).not_to be_nil
+        block_highlighted_consultations = blocks.find_by(manifest_name: :highlighted_consultations)
+        expect(block_highlighted_consultations.settings.max_results).to eq(4)
+        first_organization = Decidim::Organization.first
```
Comment (microstudi) [inline spec/system/organizations_spec.rb:68]:
```suggestion
        organization = Decidim::Organization.first
```

File: spec/system/organizations_spec.rb [hunk 6]
```diff
@@ -9,32 +9,93 @@
...
-      check "organization_available_locales_en"
-      choose "organization_default_locale_en"
-      choose "Allow participants to register and login"
       click_button "Create Organization"
 
       expect(page).to have_css("div.flash.success")
     end
+
+    context "when a real template" do
+      let(:organization) { Decidim::Organization.first }
```
Comment (microstudi) [inline spec/system/organizations_spec.rb:39]:
```suggestion
```

File: spec/system/organizations_spec.rb [hunk 7]
```diff
@@ -9,32 +9,93 @@
...
+        expect(first_organization.force_users_to_authenticate_before_access_organization).to be(false)
+        consultations = Decidim::Consultation.first
+        expect(consultations.slug).to eq("consulta-2024")
+        expect(consultations.title).to eq({ "ca" => "Consulta per Citizen Corp", "es" => "Consulta para Citizen Corp" })
+        expect(consultations.subtitle).to eq({ "ca" => "Et donem la benvinguda a la teva primera consulta!", "es" => "Te damos la bienvenida a tu primera consulta!" })
+        expect(consultations.description).to eq({ "ca" => "Et donem la benvinguda a la teva primera consulta!", "es" => "Te damos la bienvenida a tu primera consulta!" })
+        expect(consultations.banner_image.attached?).to be true
+        expect(consultations.banner_image).to be_attached
+
+        organization = Decidim::Organization.first
```
Comment (microstudi) [inline spec/system/organizations_spec.rb:81]:
```suggestion
```

File: spec/system/organizations_spec.rb [hunk 8]
```diff
@@ -9,32 +9,93 @@
...
+        expect(block_hero.settings.welcome_text).to eq({ "ca" => "Benvinguda Citizen Corp", "es" => "Bienvenida Citizen Corp" })
+        expect(block_hero.images_container.background_image.attached?).to be true
+        expect(block_hero.images_container.attached_uploader(:background_image).path).not_to be_nil
+        block_highlighted_consultations = blocks.find_by(manifest_name: :highlighted_consultations)
+        expect(block_highlighted_consultations.settings.max_results).to eq(4)
+        first_organization = Decidim::Organization.first
+        expect(first_organization.default_locale).to eq("ca")
+        expect(first_organization.users_registration_mode).to eq("enabled")
+        expect(first_organization.available_locales).to eq(%w(ca es))
+        expect(first_organization.force_users_to_authenticate_before_access_organization).to be(false)
```
Comment (microstudi) [inline spec/system/organizations_spec.rb:72]:
```suggestion
        expect(organization.default_locale).to eq("ca")
        expect(organization.users_registration_mode).to eq("enabled")
        expect(organization.available_locales).to eq(%w(ca es))
        expect(organization.force_users_to_authenticate_before_access_organization).to be(false)
```

File: app/commands/decidim/system/create_template_fields.rb [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        klass = manifest.model_class_name.constantize
+
+        params = {
+          organization: organization,
+          slug: interpolate(participatory_space["slug"]),
+          title: participatory_space["title"].transform_values { |val| interpolate(val) },
+          description: participatory_space["description"].transform_values { |val| interpolate(val) },
+          subtitle: participatory_space["subtitle"].transform_values { |val| interpolate(val) },
+          highlighted_scope: Decidim::Scope.find_by(code: participatory_space["highlighted_scope"]),
+          start_voting_date: 1.month.from_now,
```
Comment (microstudi) [inline app/commands/decidim/system/create_template_fields.rb:58]: missing video!

**Review by microstudi — COMMENTED** [review]
almost! nice job!!!

File: lib/templates/template-consultations.yml (hunk 4 above)
Comment (ElviaBth) [reply lib/templates/template-consultations.yml:220]: Done

File: app/controllers/decidim/system/custom_templates_controller.rb [hunk 10]
```diff
@@ -28,12 +28,12 @@ def create
...
 
-      def template
-        params[:template]
+      def template_id
+        params[:template_id]
       end
 
       def ensure_template_exists
-        return if template.present?
+        return if template_id.present?
```
[the remark this answers is not in the document]
Comment (ElviaBth) [reply app/controllers/decidim/system/custom_templates_controller.rb:36]: Done

File: spec/system/organizations_spec.rb (hunk 8 above)
Comment (ElviaBth) [reply spec/system/organizations_spec.rb:72]: Done

File: app/commands/decidim/system/create_template_fields.rb (hunk 9 above)
Comment (ElviaBth) [reply app/commands/decidim/system/create_template_fields.rb:58]: Ok. Thanks.

File: spec/system/organizations_spec.rb (hunk 7 above)
Comment (ElviaBth) [reply spec/system/organizations_spec.rb:81]: Done

File: spec/system/organizations_spec.rb (hunk 6 above)
Comment (ElviaBth) [reply spec/system/organizations_spec.rb:39]: Removed. Thanks

File: spec/system/organizations_spec.rb (hunk 5 above)
Comment (ElviaBth) [reply spec/system/organizations_spec.rb:68]: Done

File: lib/templates/template-consultations.yml [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        <p>Una vez en el panel de administración, modifica esta consulta para crear tu primera consulta:</p><ol>
+        <li>Desde el listado de "Consultas" haz clic en "Título de la Consulta". Accederás al formulario de edición de esta consulta.</li>
+        <li>Edita el título, el subtítulo, esta descripción y la fecha de votación adaptándolos a la información relativa a tu consulta. También puedes modificar el <em>[Alcance destacado] (cambiar literal)</em> y la imagen de cabecera, así como añadir un vídeo y/o imagen introductoria.</li>
+        <li>Haz clic en el botón "Actualización" y, a continuación, ve a la pestaña "Preguntas" que encontrarás en el menú lateral izquierdo.</li></ol>
+        <p>Si decides mantener esta consulta de prueba y prefieres crear una nueva, despublica esta para no confundir a las participantes de la plataforma. Encontrarás el botón al final de todo.</p>
+    banner_image:
+      file: template-consultations/consultation_image.png
+      content_type: image/png
```
Comment (microstudi) [inline lib/templates/template-consultations.yml:74]:
```suggestion
      content_type: image/png
      questions:
        - body:
            es: blabla
            ...
```

File: lib/templates/template-consultations.yml [hunk 12]
```diff
@@ -8,10 +8,48 @@ organization:
...
+            ca: Altres
+            es: Otros
+      - code: 1pressupost
+          name: 
+            ca: Pressupost
+            es: Presupuesto
+      - code: 0rendiciocomptes
+          name: 
+            ca: Rendició de Comptes
+            es: Rendición de Cuentas
```
Comment (microstudi) [inline lib/templates/template-consultations.yml:31]:
```suggestion
      - code: 9Altres
        name: 
          ca: Altres
          es: Otros
      - code: 1pressupost
        name: 
          ca: Pressupost
          es: Presupuesto
      - code: 0rendiciocomptes
        name: 
          ca: Rendició de Comptes
          es: Rendición de Cuentas
```
Careful, identation!

Comment (ElviaBth) [reply lib/templates/template-consultations.yml]: Thanks, done.

File: app/commands/decidim/system/create_template_fields.rb [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      questions.each do |question|
+        new_question = consultation.questions.create!(
+          title: question["title"].transform_values { |val| interpolate(val) },
+          subtitle: question["subtitle"].transform_values { |val| interpolate(val) },
+          what_is_decided: question["what_is_decided"].transform_values { |val| interpolate(val) },
+          promoter_group: question["promoter_group"],
+          question_context: question["question_context"].transform_values { |val| interpolate(val) },
+          participatory_scope: question["participatory_scope"],
+          slug: question["slug"],
+          scope: question["scope"]
```
Comment (microstudi) [inline app/commands/decidim/system/create_template_fields.rb:104]:
```suggestion
          scope:  Decidim::Scope.find_by(code: question["scope"])
```

Comment (ElviaBth) [reply app/commands/decidim/system/create_template_fields.rb]: Done

File: app/commands/decidim/system/create_template_fields.rb [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+        content_type: banner_image["content_type"],
+        metadata: nil
+      )
+    end
+
+    def create_questions(consultation, questions)
+      return unless questions
+
+      questions.each do |question|
+        new_question = consultation.questions.create!(
```
Comment (microstudi) [inline app/commands/decidim/system/create_template_fields.rb:100]:
```suggestion
        new_question = Decidim::Consultations::Question.create!(
          organization: consultation.organization,
          consultation: consultation,
```

Comment (ElviaBth) [reply app/commands/decidim/system/create_template_fields.rb:100]: Done

**Pull request merged**